proposals/003-metric-improvements.md

```diff
@@ -0,0 +1,269 @@
+| virtual_cluster | Name of virtual cluster.  Label omitted if the virtual cluster has not been established‡.                                                                                 |
+| node_id         | Numeric node_id of the broker being communicated with. the value `bootstrap` is used for bootstrapping interactions. Label omitted if the node has not been established‡. |
+| api_key         | API key of the kafka message e.g. PRODUCE                                                                                                                                 |                                                                                                                                               |
+| api_version     | Numeric API version of the message.                                                                                                                                       |                                                                                                                                               |
+| decoded         | Flag indicating if the proxy decoded the message or not (1 = decoded, 0 = not decoded)                                                                                    |                                                                                                                                               |
+| filter_name     | Name of the filter that initiated the request                                                                                                                             |                                                                                                                                               |
+
+The following metrics will be deprecated.
+
+| Metric                                           |
+|--------------------------------------------------|
+| kroxylicious_inbound_downstream_messages         |
+| kroxylicious_inbound_downstream_decoded_messages |
+| kroxylicious_payload_size_bytes                  |
+
+† Requests generated by Filter API's `sendRequest`, and their responses, are not accounted for by these metric
+
+‡ This happens only in the case of connection to gateways using "SNI identifies broker" where the connection fails TLS negotiation or the SNI name is unrecognised.
+
+### New counters `kroxylicious_<A>_to_<B>_request_total` and `kroxylicious_<A>_to_<B>_response_total`
+
+Counts the number of request/responses flowing between point A and point B.
+
+Use cases that were previously served by the deprecated metrics `kroxylicious_inbound_downstream_messages` and
+`kroxylicious_inbound_downstream_decoded_messages` can now use these metric instead.
+
+In the case where a filter **short-circuits** a request and provides its own response, `client_to_proxy_request_total` will tick
+but `kroxylicious_proxy_to_broker_request_total` will not.
```

Comment:
I wonder if we should consider a specific counter, or `short_circuit` label on `kroxylicious_proxy_to_client_response_total`. Something users can look at to convince themselves this accounts for the discrepancy in and out.

Reply:
I've recorded this in the out-of-scope section.